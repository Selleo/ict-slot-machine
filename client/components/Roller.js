import React from 'react'
import ReactDOM from 'react-dom'
import times from 'lodash/times'
import random from 'lodash/random'

import Slot from './Slot'
import Animation from './Animation'
import Polygon, { getR } from '../modules/polygon'

export default class Roller extends React.Component {
  constructor(props) {
    super(props)

    this.polygon = new Polygon(props.height, props.count)
    this.spin = this.spin.bind(this)
  }

  spin(startingIndex = 0, forcedSpinTo) {
    return new Promise(resolve => {
      const { count, position } = this.props

      let rotations = random(30, 50)
      const animationTime = random(3, 6, true)

      if (isFinite(forcedSpinTo)) {
        rotations = 3 * count + forcedSpinTo - startingIndex
      }

      setTimeout(() => {
        resolve({ [position]: (startingIndex + rotations) % count })
      }, animationTime * 1000 + 500)

      ReactDOM.render(
        <Animation
          startingIndex={startingIndex}
          position={position}
          animationTime={animationTime}
          rotations={rotations}
          angle={this.polygon.angle}
          offset={5}
        />,
        document.querySelector(`.styles-roller-${position}`)
      )
    })
  }

  _createSlots() {
    return times(this.props.count, number => {
      return <Slot key={number} polygon={this.polygon} number={number} />
    })
  }

  _getRollerStyles() {
    const { width, height, count } = this.props

    return {
      transformOrigin: `0 50% -${getR(height, count)}px`,
      width: `${width}px`,
      height: `${height}px`
    }
  }

  render() {
    const { position } = this.props

    return (
      <div className={`roller-wrapper ${position}`}>
        <div className={`styles-roller-${position}`} />

        <div className={`roller ${position}`} style={this._getRollerStyles()}>
          {this._createSlots()}
        </div>
      </div>
    )
  }
}
