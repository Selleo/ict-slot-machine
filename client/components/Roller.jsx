import React from 'react'
import ReactDOM from 'react-dom'
import Slot from './Slot'
import Animation from './Animation'
import Polygon, { getR } from '../modules/polygon'
import { times, random } from 'lodash'

export default class Roller extends React.Component {
  constructor(props) {
    super(props)

    this.polygon = new Polygon(props.height, props.count)
    this.spin = this.spin.bind(this)
  }

  spin(startingIndex = 0) {
    return new Promise(resolve => {
      const { count, position } = this.props

      const rotations = random(30, 50)
      const animationTime = random(3, 5, true)

      setTimeout(() => {
        resolve({ [position]: (startingIndex + rotations) % count })
      }, animationTime * 1000 + 500)

      ReactDOM.render(
        <Animation
          startingIndex={startingIndex}
          position={position}
          animationTime={animationTime}
          rotations={rotations}
          angle={this.polygon.angle}
          offset={5}
        />,
        document.querySelector(`.styles-roller-${position}`)
      )
    })
  }

  _createSlots() {
    return times(this.props.count, number => {
      return <Slot key={number} polygon={this.polygon} number={number} />
    })
  }

  _getRollerStyles() {
    const { width, height, count } = this.props

    return {
      transformOrigin: `0 50% -${getR(height, count)}px`,
      width: `${width}px`,
      height: `${height}px`
    }
  }

  _getRollerWrapperStyles(position) {
    return {
      perspectiveOrigin: `${HORIZONTAL_PERSPECTIVE[position]}% 50%`
    }
  }

  render() {
    const { position } = this.props

    return (
      <div className={`roller-wrapper ${position}`}>
        <div className={`styles-roller-${position}`} />

        <div className={`roller ${position}`} style={this._getRollerStyles()}>
          {this._createSlots()}
        </div>
      </div>
    )
  }
}
