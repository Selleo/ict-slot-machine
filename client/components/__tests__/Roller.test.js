import React from 'react'
import renderer from 'react-test-renderer'
import Roller from '../Roller'

describe('<Roller />', () => {
  const defaultProps = {
    position: 'left',
    width: 100,
    height: 100,
    count: 10
  }

  const wrapper = renderer.create(<Roller {...defaultProps} />)

  test('render', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
