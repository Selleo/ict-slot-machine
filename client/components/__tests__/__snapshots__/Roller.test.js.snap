// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Roller /> render 1`] = `
<div
  className="roller-wrapper left"
>
  <div
    className="styles-roller-left"
  />
  <div
    className="roller left"
    style={
      Object {
        "height": "100px",
        "transformOrigin": "0 50% -153.8841768587627px",
        "width": "100px",
      }
    }
  >
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(0px) translateY(0px) rotateX(0deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-29.38926261462366px) translateY(90.45084971874738px) rotateX(-36deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-106.331351044005px) translateY(146.35254915624213px) rotateX(-72deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-201.43700267352034px) translateY(146.35254915624213px) rotateX(-108deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-278.3790911029017px) translateY(90.45084971874739px) rotateX(-144deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-307.7683537175254px) translateY(1.884537646295089e-14px) rotateX(-180deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-278.3790911029017px) translateY(-90.45084971874736px) rotateX(-216deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-201.4370026735204px) translateY(-146.35254915624213px) rotateX(-252deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-106.33135104400503px) translateY(-146.35254915624213px) rotateX(-288deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="slot"
      style={
        Object {
          "transform": "translateZ(-29.389262614623675px) translateY(-90.45084971874742px) rotateX(-324deg)",
        }
      }
    >
      <img
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;
